src/collapse/mod.rs

```diff
@@ -23,9 +23,17 @@ use std::sync::Arc;
 
 use chashmap::CHashMap;
 use fnv::FnvHashMap;
+use lazy_static::lazy_static;
+
+lazy_static! {
+    #[doc(hidden)]
+    pub static ref DEFAULT_NTHREADS: usize = num_cpus::get();
+}
 
 const CAPACITY_HASHMAP: usize = 512;
 const CAPACITY_READER: usize = 128 * 1024;
+#[doc(hidden)]
+pub const DEFAULT_NSTACKS: usize = 20;
```

Comment:
I'd love to see some benchmarks on the effects of tuning this parameter.

Reply:
It's there! Look in the tests section of `collapse::perf`

Reply:
`cargo test vary_nstacks --release -- --ignored --nocapture`

Reply:
Ah, interesting, missed that! Here's what I get on a 40-core machine:
```
./flamegraph/example-perf-stacks.txt.gz (nstacks: 1316, lines: 37158)
    nstacks_per_job:  705 (% of total:  54%) | time: 98% of default
    nstacks_per_job:  724 (% of total:  55%) | time: 98% of default
    nstacks_per_job:  735 (% of total:  56%) | time: 98% of default
    nstacks_per_job:  813 (% of total:  62%) | time: 98% of default
    nstacks_per_job:  848 (% of total:  64%) | time: 98% of default
    nstacks_per_job:  765 (% of total:  58%) | time: 98% of default
    nstacks_per_job:  676 (% of total:  51%) | time: 98% of default
    nstacks_per_job:  498 (% of total:  38%) | time: 98% of default
    nstacks_per_job:  882 (% of total:  67%) | time: 98% of default
    nstacks_per_job:  567 (% of total:  43%) | time: 98% of default

./flamegraph/test/perf-vertx-stacks-01.txt (nstacks: 286, lines: 9985)
    nstacks_per_job:  112 (% of total:  39%) | time: 99% of default
    nstacks_per_job:  157 (% of total:  55%) | time: 99% of default
    nstacks_per_job:  273 (% of total:  95%) | time: 99% of default
    nstacks_per_job:   67 (% of total:  23%) | time: 99% of default
    nstacks_per_job:  122 (% of total:  43%) | time: 99% of default
    nstacks_per_job:   14 (% of total:   5%) | time: 99% of default
    nstacks_per_job:  204 (% of total:  71%) | time: 99% of default
    nstacks_per_job:  154 (% of total:  54%) | time: 99% of default
    nstacks_per_job:  189 (% of total:  66%) | time: 99% of default
    nstacks_per_job:  230 (% of total:  80%) | time: 99% of default

./flamegraph/test/perf-iperf-stacks-pidtid-01.txt (nstacks: 201, lines: 3608)
    nstacks_per_job:  149 (% of total:  74%) | time: 98% of default
    nstacks_per_job:   95 (% of total:  47%) | time: 98% of default
    nstacks_per_job:   53 (% of total:  26%) | time: 98% of default
    nstacks_per_job:  186 (% of total:  93%) | time: 98% of default
    nstacks_per_job:  172 (% of total:  86%) | time: 98% of default
    nstacks_per_job:   44 (% of total:  22%) | time: 98% of default
    nstacks_per_job:   30 (% of total:  15%) | time: 98% of default
    nstacks_per_job:  188 (% of total:  94%) | time: 98% of default
    nstacks_per_job:   37 (% of total:  18%) | time: 98% of default
    nstacks_per_job:  103 (% of total:  51%) | time: 98% of default

./flamegraph/test/perf-numa-stacks-01.txt (nstacks: 200, lines: 2762)
    nstacks_per_job:  193 (% of total:  96%) | time: 99% of default
    nstacks_per_job:  153 (% of total:  76%) | time: 100% of default
    nstacks_per_job:    6 (% of total:   3%) | time: 100% of default
    nstacks_per_job:   63 (% of total:  32%) | time: 100% of default
    nstacks_per_job:  174 (% of total:  87%) | time: 100% of default
    nstacks_per_job:  137 (% of total:  68%) | time: 100% of default
    nstacks_per_job:  103 (% of total:  52%) | time: 100% of default
    nstacks_per_job:   92 (% of total:  46%) | time: 100% of default
    nstacks_per_job:  165 (% of total:  82%) | time: 100% of default
    nstacks_per_job:   61 (% of total:  30%) | time: 100% of default
```
Any particular reason you chose 20 as the default?

Reply:
The `time: X% of default` is what matters most. That's the time each run took as a % of the time running with `nstacks_per_job = 20` takes. As you can see, any number you pick (on your machine at least) performs roughly the same (getting number like `98%` means there's very little difference between that run and the default). As I looked at things on my machine, I recall around 20 being a sweet spot, but it also didn't matter that much for me.

Reply:
I'm not quite done tweaking the code. Once I'm completely finished I'll run this again and see if I can make any sense out of it. My guess is, though, it's not going to matter much.

Reply:
Turns out there was a bug in that code, which I've now fixed. It wasn't varying `nstacks_per_job`!!!, which definitely explains why everything was showing up the same before. Oops :). 

I'd be curious to see what you get now when you run it again. The command is now `cargo test bench_nstacks_perf --release -- --ignored --nocapture`. You can also now run `cargo test bench_nstacks_dtrace --release -- --ignored --nocapture` for `dtrace`.

100 seems to be a sweat spot for me for both `perf` and `dtrace`, but what's optimal is probably a function of all kinds of things (like number of cores); so it's hard to know a) what we should use as a default, and b) how to communicate to the user (if we choose to keep it as a parameter he/she can vary) valuable heuristics for how to set it.

Reply:
100 seems closer to what I'd expect this value to be :)
I'll run it now and see what I get!

Reply:
```

Reply:
```
./flamegraph/example-perf-stacks.txt.gz (nstacks: 1316, lines: 37158)
    nstacks_per_job:   70 (% of total:   5%) | time: 99% of default
    nstacks_per_job:   80 (% of total:   6%) | time: 99% of default
    nstacks_per_job:   90 (% of total:   7%) | time: 100% of default
    nstacks_per_job:  100 (% of total:   8%) | time: 100% of default
    nstacks_per_job:  140 (% of total:  11%) | time: 100% of default
    nstacks_per_job:  130 (% of total:  10%) | time: 100% of default
    nstacks_per_job:  110 (% of total:   8%) | time: 100% of default
    nstacks_per_job:  170 (% of total:  13%) | time: 101% of default
    nstacks_per_job:  200 (% of total:  15%) | time: 101% of default
    nstacks_per_job:   40 (% of total:   3%) | time: 101% of default

./flamegraph/test/perf-vertx-stacks-01.txt (nstacks: 286, lines: 9985)
    nstacks_per_job:   10 (% of total:   3%) | time: 62% of default
    nstacks_per_job:   20 (% of total:   7%) | time: 67% of default
    nstacks_per_job:   30 (% of total:  10%) | time: 73% of default
    nstacks_per_job:   40 (% of total:  14%) | time: 75% of default
    nstacks_per_job:   50 (% of total:  17%) | time: 76% of default
    nstacks_per_job:    1 (% of total:   0%) | time: 76% of default
    nstacks_per_job:   60 (% of total:  21%) | time: 79% of default
    nstacks_per_job:   70 (% of total:  24%) | time: 86% of default
    nstacks_per_job:   80 (% of total:  28%) | time: 91% of default
    nstacks_per_job:   90 (% of total:  31%) | time: 95% of default

./flamegraph/test/perf-iperf-stacks-pidtid-01.txt (nstacks: 201, lines: 3608)
    nstacks_per_job:   10 (% of total:   5%) | time: 74% of default
    nstacks_per_job:   20 (% of total:  10%) | time: 76% of default
    nstacks_per_job:   30 (% of total:  15%) | time: 77% of default
    nstacks_per_job:   40 (% of total:  20%) | time: 82% of default
    nstacks_per_job:   50 (% of total:  25%) | time: 84% of default
    nstacks_per_job:   70 (% of total:  35%) | time: 87% of default
    nstacks_per_job:   60 (% of total:  30%) | time: 87% of default
    nstacks_per_job:   80 (% of total:  40%) | time: 91% of default
    nstacks_per_job:    1 (% of total:   0%) | time: 93% of default
    nstacks_per_job:   90 (% of total:  45%) | time: 96% of default

./flamegraph/test/perf-numa-stacks-01.txt (nstacks: 200, lines: 2762)
    nstacks_per_job:   10 (% of total:   5%) | time: 80% of default
    nstacks_per_job:   30 (% of total:  15%) | time: 82% of default
    nstacks_per_job:   20 (% of total:  10%) | time: 82% of default
    nstacks_per_job:   40 (% of total:  20%) | time: 84% of default
    nstacks_per_job:   50 (% of total:  25%) | time: 86% of default
    nstacks_per_job:   60 (% of total:  30%) | time: 89% of default
    nstacks_per_job:   70 (% of total:  35%) | time: 91% of default
    nstacks_per_job:   80 (% of total:  40%) | time: 94% of default
    nstacks_per_job:   90 (% of total:  45%) | time: 98% of default
    nstacks_per_job:  100 (% of total:  50%) | time: 100% of default
```
```
./flamegraph/example-dtrace-stacks.txt (nstacks: 3413, lines: 41913)
    nstacks_per_job:   60 (% of total:   2%) | time: 98% of default
    nstacks_per_job:  160 (% of total:   5%) | time: 99% of default
    nstacks_per_job:   90 (% of total:   3%) | time: 99% of default
    nstacks_per_job:   80 (% of total:   2%) | time: 100% of default
    nstacks_per_job:   50 (% of total:   1%) | time: 100% of default
    nstacks_per_job:  140 (% of total:   4%) | time: 100% of default
    nstacks_per_job:  150 (% of total:   4%) | time: 100% of default
    nstacks_per_job:  100 (% of total:   3%) | time: 100% of default
    nstacks_per_job:  120 (% of total:   4%) | time: 100% of default
    nstacks_per_job:  180 (% of total:   5%) | time: 100% of default
```

Reply:
I think 100 is a reasonable default value.